### ISSUE TITLE
The Need for Super-Organizers

### ISSUE THREAD

**Issue by adomokos** [adomokos/light-service#65]
As our business requirements and code complexity grows, we see the need of weaving together Organizers just like we do with Actions. We have conventions: an Action can not call other Action(s), an Organizer can not invoke other Organizer(s). This is how we try to keep our code simple and easy to follow.

We have been calling multiple Organizers from controller actions which is a nice way of pushing out logic from Rails, but it seems it's not enough. We could build new organizers that reuse all the actions from the other organizers, but I am not sure that's the best solution, as the Organizer is responsible for one specific logic. For example, one Organizer handles Payment, the other communicates to an external API if the Payment was successful.

We've been considering "Orchestrators" or "Super Organizers", that would treat Organizers as higher level Actions. With this change the Organizers have a uniform interface, they might express `expects` and `promises` as well.

Do you have any thoughts on this?

**Comment by mattr-** [comment]
:+1: to everything here.

I like the term "Orchestrators" for these types of objects, myself.

**Comment by adomokos** [comment]
Thank you, @mattr-! Good to hear from you :wink:!

**Comment by padi** [comment]
@adomokos we've been using classes that are both `include`-ing or `extend`-ing `Organizer` and `Action`, how will this be different?

**Comment by adomokos** [comment]
An `Orchestrator` will "orchestrate" Organizers. When you see an Orchestrator, you know that its responsibility is mediating between Organizers.

This is what we can do with Organizers:
```
Organizer1
  |- Action1
  |- Action2
```

And this is what we will able to do with Orchestrators:
```
Orchestrator
  |- Organizer1
  |     |- Action1
  |     |- Action2
  |- Organizer2
        |-Action3
        |-Action4
```

Orchestrator simply sets up the initial Context and reduces the Organizers. Not sure how this will look, but the similarity between Organizer and Orchestrator is close. We just need to spike this out a bit to see what it will look - and feel - like.

@padi: Is this description helping? Let me know if you have further questions.

**Comment by padi** [comment]
Historically, we've been using this format for quite some time now:

```
Organizer (include
  |- OrganizerAction1*
  |     |- Action1
  |     |- Action2
  |- OrganizerAction2*
        |-Action3
        |-Action4
```

`OrganizerAction1` and `OrganizerAction2` are classes that include both `Organizer` and `Action` e.g.

```ruby
class Organizer1
  include LightService::Action
  include LightService::Organizer

  executed do |ctx|
    with(ctx).reduce([Action1, Action2])
  end
end
```

**Comment by adomokos** [comment]
I'd assume `Organizer1` is your `OrganizerAction1*`, according to your tree structure above.

I like what you did there, maybe that's what we should consider as well.

I guess the "Orchestrator" would make this universal, so people don't have to reinvent the wheel. Look at me, I have not even thought about including both the `Organizer` and `Action` in one class.

**Comment by padi** [comment]
@adomokos Yes. I changed `Organizer1` to `OrganizerAction1` now.

**Comment by pzula** [comment]
@adomokos Do we want to try to include super organizers as a spike?

**Comment by adomokos** [comment]
@pzula: Thank you for pinging me about it! At my previous job I had a spike for this. It used Ruby lambdas heavily. I'll try to create a spike branch based on those ideas and then we have something we could iterate on. If you have other ideas, I'd love to see them in spike branch!

**Comment by adomokos** [comment]
As we are trying to build more complex ETLs with LightService, we started narrowing down the scope of "orchestrators".

Here are the goals that I think is needed for a PR to close this issue:

1. Regardless of complexity, we should be able to map the entire workflow in 1 organizer
2. We need to handle iterators
3. We need to handle conditionals

<b>1. One workflow</b>
As complexity grows, we started creating child organizers with its own set of actions. We put them in separate files. When we wanted to check what's happening where, we had to open the spaghetti of organizers. Even worse, we wrapped an organizer call into an action as we iterated through a collection and called an organizer with each item.

<b>2. Handle Iterators</b>
I described that above already, we had to iterate over a collection and call a child organizer to deal with it. Big no-no.

<b>3. Handle Conditionals</b>
Depending on the state of the items in the context, I may or may not want to call and action or a set of actions from an organizer. Obviously I could use an action to make that decision, but that check is then hidden from the organizer. I have to dig to find that information.

Here is what I came up with as a starting point:

```ruby
module ETL
  class RunsProcess
    extend LightService::Organizer

    def self.call(run_id)
      with(:run_id => run_id)
        .reduce(actions)
    end

    def self.actions
      [
        Action1,
        Action2,
        Action3,
        Class.new do
          extend LightService::Action
          extend LightService::Organizer
          expects :run_id, :items

          executed do |ctx|
            ctx.items.each do |item|
              with(:run_id => ctx.run_id,
                     :item => item)
                .reduce([
                          SubAction1,
                          SubAction2,
                          SubAction3,
                          SubAction4])
            end
          end
        end
      ]
    end
  end
end
```

This works, but it's very noisy. It's hard to see the what's really going on here.

This was proposed to make it better:
```ruby
module ETL
  class RunsProcess
    ...
    def self.actions
      [
        Action1,
        Action2,
        Action3,
        iterate(:documents).over([
            SubAction1,
            SubAction2,
            SubAction3,
            SubAction4])
      ]
    end
  end
end
```

This is much better, and I think it addresses point 1 and 2 above. I think we would build this and add it to LS if others think it's a good idea.

But what are we doing with the conditionals?

Would something like this work?
```ruby
module ETL
  class RunsProcess
    ...
    def self.actions
      [
        Action1,
        Action2,
        Action3,
        iterate(:documents).over([
            SubAction1,
            SubAction2,
            SubAction3,
            SubAction4]),
        condition(-> { ctx.something.count == 2 }).with([
            SubAction5,
            SubAction6,
            SubAction7
        ]),
        Action4,
        Action5
      ]
    end
  end
end
```
What do people think about this? The lambda encapsulates the condition and it's invoked when that particular step is being processed.

Or, we could just create an action that acts like a guard for a series of child actions like this:
```ruby
module ETL
  class RunsProcess
    ...
    def self.actions
      [
        Action1,
        Action2,
        Action3,
        iterate(:documents).over([
            SubAction1,
            SubAction2,
            SubAction3,
            SubAction4]),
        [SkipThisIfBooleanLogicIsFalseAction,
         SubAction5,
         SubAction6,
         SubAction7],
        Action4,
        Action5
      ]
    end
  end
end
```

Any feedback, comment is much appreciated.

**Comment by adomokos** [comment]
Btw, this is the last remaining piece before I'd bump LightService to 1.0.

**Comment by mattr-** [comment]
I like the specification for iteration and conditionals based on the context state, so :+1: for that.

I dislike the connascence of position implied with the following code though.
```ruby
[SkipThisIfBooleanLogicIsFalseAction,
 SubAction5,
 SubAction6,
 SubAction7],
```

Would it be possible to use a hash instead? That would give us something like the following:
```ruby
{
  SkipThisIfBooleanLogicIsFalseAction: [
    SubAction5,
    SubAction6,
    SubAction7,
  ]
}
```

I think using a hash gives you a clearer division between the condition and the operations to perform should the condition match and switches us to connascence of meaning, which is a weaker (and therefore easier to break) form of connascence.

All of this does assume that the second form was intended as a type of shorthand for the `condition` clause just above it. 

My :two: :moneybag: worth.

**Comment by padi** [comment]
@adomokos here's my suggestion, taking a bit from all of your suggestions above.

```ruby
module ETL
  class RunsProcess
    ...
    def self.actions
      [
        Action1,
        Action2,
        Action3,
       # just in case you want to manipulate more than 1 variable
        iterate(:documents, :other_files, if: -> { ctx.something.count == 2 }).over([
            SubAction1,
            SubAction2,
            SubAction3,
            SubAction4
        ])
      ]
    end
  end
end
```

... or ...

```ruby
module ETL
  class RunsProcess
    ...
    def self.actions
      [
        Action1,
        Action2,
        Action3,
       # just in case you want to manipulate more than 1 variable
        iterate(:documents, :other_files, if: SkipThisIfBooleanLogicIsFalseAction).over([
            SubAction1,
            SubAction2,
            SubAction3,
            SubAction4
        ])
      ]
    end
  end
end
```

The `iterate`'s `if` option can accept a lambda (that is `instance_eval`'d) or a service object (possibly a `LightService::Action` that will evaluate whether or not the sub-actions will be executed.

**Comment by adomokos** [comment]
@mattr-, thank you for your suggestion. I like the special position the first action has in your comment.

@padi: why should we combine the `iterate` method with a conditional? In the example I proposed, I see no connection between the two (iterators and conditionals). Also, what is `:other_files`? Can you please provide more context?

**Comment by padi** [comment]
@adomokos I thought `iterate` meant it was going to iterate any set of context key-values over a series of actions, hence the `:other_files`. If `iterate` was only meant for strictly one context key-value, then I'm mistaken.

As for `condition(lambda).with()`, I'm not fond of using `with` for 2 reasons: 

1. If we're going for DSLs, then it must at least make sense when read in plain English. Iterating a something over a set of sub-actions seems to adequately readable to me, so `iterate(:something).over(...)` is ok. 

2. Method overloading. `with` has been previously used in `with(...).reduce(...)`. Even if we disregard 1, it seems that the more consistent option would be `condition(...).reduce(...)`, since `reduce` has been accepting an array of sub-actions

**Comment by adomokos** [comment]
We've been working on the "orchestrator" part of LS in the last couple of months.
Check out [the work](https://github.com/adomokos/light-service/compare/master...KennaSecurity:orchestrators?expand=1) we have there, I'll extract those into several PRs in the coming weeks.

**Comment by adomokos** [comment]
I started porting over the `Orchestrator` specific code. You can find it in [this branch](https://github.com/adomokos/light-service/tree/ad-orchestrators#orchestrators). We will use this in production in our projects, and we'll see how it feels.

I'll also open a PR, so people can review it, suggest better names, simpler solutions.

**Comment by adomokos** [comment]
Closing this, as PR #91 has been merged.
